File: commerce/core/categories.py
"""Shop categories — the canonical taxonomy (§8 trending rail + feed card colors).

A shop's CATEGORY is what gives the trending rail its different-stimulus colors (a butchery reads
red, a bakery amber, …). The taxonomy lives here as the single backend source of truth: the API
validates a shop's declared category against this allow-list (an unknown value is a 422 at the
edge — never free-text into the rail, S-input-validation), and the feed/profile/trending DTOs
carry only the opaque SLUG.

The COLOR is deliberately NOT stored or returned by the backend — it is a pure presentation
concern derived client-side from the slug (frontend ``utils/categories.ts`` mirrors this list and
maps each slug to a CSS color token). Keeping the slug→color map out of the API means a palette
re-tune is a frontend-only change and the wire stays a stable, minimal contract.

A shop with NO category (None) is legal — an un-categorised / personal-timeline shop. The rail
falls back to a neutral treatment for those; only a declared, valid slug gets a category color.
"""
from __future__ import annotations

# The legal category slugs. Order is presentational-irrelevant (the frontend owns ordering of the
# picker), but kept stable so the parity test against the frontend list is deterministic.
SHOP_CATEGORIES: tuple[str, ...] = (
    "butchery",
    "bakery",
    "greengrocer",
    "restaurant",
    "boutique",
    "electronics",
    "shoes",
    "beauty",
    "hardware",
    "pharmacy",
    "general",
)

# A set for O(1) membership in the hot validation path.
_CATEGORY_SET = frozenset(SHOP_CATEGORIES)


def is_valid_category(value: str | None) -> bool:
    """True iff ``value`` is a known category slug. ``None`` (unset) is accepted by the schema
    separately — this guards only that a SUPPLIED category is one we recognise."""
    return value in _CATEGORY_SET

File: weespas-frontend/src/utils/categories.ts
// Shop categories — the frontend mirror of the backend taxonomy (commerce core/categories.py).
//
// The BACKEND owns the legal slugs and validates them; it returns only the opaque slug on the wire
// (FeedItem.shop_category / ShopProfile.category). The COLOR is a purely-presentational concern
// owned here: each slug maps to a human label (for the seller's category picker) and a CSS color
// token (defined in styles/variables.css) that paints the §8 trending rail's category cards and a
// subtle tint on the feed card. This split means a palette re-tune never touches the API.
//
// A parity test (categories.test.ts) asserts CATEGORY_SLUGS here stays in lock-step with the
// backend list, so the two can't silently drift (an unknown slug would otherwise render as the
// neutral fallback with no error).

export interface CategoryMeta {
  /** Human label for the seller's category picker + the rail card. */
  readonly label: string;
  /** The CSS custom property (defined in variables.css) carrying this category's color. */
  readonly colorVar: string;
}

// slug → { label, colorVar }. Order here drives the seller picker's option order.
export const CATEGORY_META: Record<string, CategoryMeta> = {
  butchery: { label: 'Butchery', colorVar: '--color-cat-butchery' },
  bakery: { label: 'Bakery', colorVar: '--color-cat-bakery' },
  greengrocer: { label: 'Mama Mboga / Greengrocer', colorVar: '--color-cat-greengrocer' },
  restaurant: { label: 'Food & Restaurant', colorVar: '--color-cat-restaurant' },
  boutique: { label: 'Boutique / Exotic / Wedding', colorVar: '--color-cat-boutique' },
  electronics: { label: 'Electronics', colorVar: '--color-cat-electronics' },
  shoes: { label: 'Shoe Store', colorVar: '--color-cat-shoes' },
  beauty: { label: 'Beauty & Salon', colorVar: '--color-cat-beauty' },
  hardware: { label: 'Hardware', colorVar: '--color-cat-hardware' },
  pharmacy: { label: 'Pharmacy', colorVar: '--color-cat-pharmacy' },
  general: { label: 'General / Duka', colorVar: '--color-cat-general' },
};

/** The legal category slugs, in picker order. Kept in parity with the backend SHOP_CATEGORIES. */
export const CATEGORY_SLUGS: readonly string[] = Object.keys(CATEGORY_META);

/** A neutral color token for an un-categorised shop (null/unknown slug) — keeps the rail coherent
 *  without implying a category. Defined in variables.css. */
export const CATEGORY_NEUTRAL_VAR = '--color-cat-neutral';

/** The CSS `var(...)` color expression for a category slug (or the neutral fallback for
 *  null/unknown). Use directly in an inline style, e.g. `style={{ color: categoryColor(c) }}`. */
export function categoryColor(slug: string | null | undefined): string {
  const meta = slug ? CATEGORY_META[slug] : undefined;
  return `var(${meta ? meta.colorVar : CATEGORY_NEUTRAL_VAR})`;
}

/** The display label for a category slug; '' for null/unknown (the caller omits the chip). */
export function categoryLabel(slug: string | null | undefined): string {
  const meta = slug ? CATEGORY_META[slug] : undefined;
  return meta ? meta.label : '';
}
